csaf-import/src/main/kotlin/io/github/csaf/sbom/Retrieval.kt

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright (c) 2024, The Authors. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+package io.github.csaf.sbom
+
+import io.github.csaf.sbom.generated.Provider
+import io.ktor.client.HttpClient
+import io.ktor.client.request.*
+
+/**
+ * Retrieves one or more provider-metadata.json documents (represented by the [Provider] data class)
+ * from a domain according to the retrieval rules defined in
+ * https://docs.oasis-open.org/csaf/csaf/v2.0/os/csaf-v2.0-os.html#731-finding-provider-metadatajson.
+ */
+suspend fun retrieveProviderFromDomain(domain: String): Result<List<Provider>> {
```

Comment:
Why is that a list of `Provider`s? Thought that only aggregators and listers return lists of Providers?

Reply:
You can also discover providers by specifying a domain. See https://docs.oasis-open.org/csaf/csaf/v2.0/os/csaf-v2.0-os.html#731-finding-provider-metadatajson. Because the provider metadata can be found in multiple places, we potentially need to return a list of providers (or at least of the provider-metadata).

As mentioned in the other PR it could probably make sense to separate the logical entity of a "provider" (then in this case it would be only one) but it can possibly have multiple "provider metadata" (the JSON). Unfortunately, the specification only says: "Select one or more provider-metadata.json to use.". So we would probably also free to just use the first one for example.